pkg/compute/bidder.go

```diff
@@ -47,110 +53,160 @@ func NewBidder(params BidderParams) Bidder {
+	calculatedResources *models.Resources
+}
 
-	// if the request is pre-approved, we should run the execution immediately if the bid is accepted
-	if !request.WaitForApproval {
-		if !response.ShouldBid || response.ShouldWait {
+func (b Bidder) handleBidResult(
+	ctx context.Context,
+	result *bidStrategyResponse,
+	targetPeer string,
+	waitForApproval bool,
+	execution *models.Execution,
+) {
+	var (
+		routingMetadata = RoutingMetadata{
+			// the source of this response is the bidders nodeID.
+			SourcePeerID: b.nodeID,
+			// the target of this response is the source of the request.
+			TargetPeerID: targetPeer,
+		}
+		executionMetadata = ExecutionMetadata{
+			ExecutionID: execution.ID,
+			JobID:       execution.JobID,
+		}
+		handleComputeFailure = func(ctx context.Context, err error, reason string) {
+			var errMsg string
+			if err != nil {
+				log.Ctx(ctx).Error().Err(err).Msg(reason)
+				errMsg = fmt.Sprintf("%s: %s", reason, err)
+			} else {
+				log.Ctx(ctx).Info().Msg(reason)
+				errMsg = reason
+			}
 			b.callback.OnComputeFailure(ctx, ComputeError{
 				RoutingMetadata:   routingMetadata,
 				ExecutionMetadata: executionMetadata,
-				Err:               fmt.Sprintf("Job rejected: %s", response.Reason),
+				Err:               errMsg,
 			})
-		} else {
-			execution := store.NewLocalExecutionState(request.Execution, request.SourcePeerID)
-			execution.State = store.ExecutionStateBidAccepted
-			if err := b.store.CreateExecution(ctx, *execution); err != nil {
-				log.Ctx(ctx).Error().Err(err).Msg("Unable to create execution state")
-				return
-			}
-			err := b.executor.Run(ctx, *execution)
-			if err != nil {
-				log.Ctx(ctx).Error().Err(err).Msg("Unable to run execution")
-				return
-			}
 		}
-		return
-	}
+		handleBidComplete = func(ctx context.Context, result *bidStrategyResponse) {
+			b.callback.OnBidComplete(ctx, BidResult{
+				RoutingMetadata:   routingMetadata,
+				ExecutionMetadata: executionMetadata,
+				Accepted:          result.bid,
+				Wait:              result.wait,
+				Reason:            result.reason,
+			})
+		}
+	)
 
-	bidResult := BidResult{
-		RoutingMetadata:   routingMetadata,
-		ExecutionMetadata: executionMetadata,
-		Accepted:          response.ShouldBid,
-		Reason:            response.Reason,
+	if !waitForApproval {
+		if !result.bid || result.wait {
+			handleComputeFailure(ctx, nil, fmt.Sprintf("job rejected: %s", result.reason))
+			return
+		}
+
+		execution.AllocateResources(execution.Job.Task().Name, *result.calculatedResources)
+		localExecution := store.NewLocalExecutionState(execution, targetPeer)
+		localExecution.State = store.ExecutionStateBidAccepted
+
+		if err := b.store.CreateExecution(ctx, *localExecution); err != nil {
+			handleComputeFailure(ctx, err, "failed to create execution state")
+			return
+		}
+		if err := b.executor.Run(ctx, *localExecution); err != nil {
+			handleComputeFailure(ctx, err, "failed to run execution")
+			return
+		}
```

Comment:
In addition to this component handling the compute failure by calling OnComputeFailure when Run returns an error, the executor_buffer (which this executor is) also has this in its Run method.

```
	defer func() {
		if err != nil {
			s.callback.OnComputeFailure(ctx, ComputeError{
				ExecutionMetadata: NewExecutionMetadata(execution),
				RoutingMetadata: RoutingMetadata{
					SourcePeerID: s.ID,
					TargetPeerID: localExecutionState.RequesterNodeID,
				},
				Err: err.Error(),
			})
		}
	}()
```

This feels like it might be the cause of the race when we occassionally stop a job but it's in the wrong state.

Reply:
oooo great catch!
70c9bdc6d21122340f7369789a45a4937e303dad